request-response-interface/README.md

```diff
@@ -0,0 +1,107 @@
+
+Many application protocols map well to request/response semantics where a peer
+makes a request to another peer, and receives a response back. For example,
+requesting a file from a peer and getting the file back. In contrast to
+request/response, stream semantics allow two peers to continously send data
+back and forth with no specific ordering. Both styles have their uses. This
+document focuses on _one_ way of implementing request/response semantics and
+how to map it to stream based transports and an HTTP transport.
+
+## The interface
+
+At its core, request-response means:
+- taking _one_ blob from the requesting peer (aka the client)
+- delivering this blob to the responding peer (aka the server)
+- taking _one_ blob from the responding peer
+- delivering this blob to the requesting peer
+
+The defining characteristics are:
+
+- Each party can only send one message
+- The interaction is linear in time (the response can only be initiated after the request has been received)
+
+This is in contrast to for example streaming protocols where the responding peer may send any number of messages or event/notification-based protocols where a message may be received without a prior request.
+
+What the blobs are and how they are encoded is subject to the application
+protocol and not defined in this interface.
+
+A rough suggestion for implementors to follow is provide something like:
+```
+async fn handleRequest(request) -> response
+```
+
+where the request/response may be read incrementally and asynchronously and the
+request/response may also be sent incrementally and asynchronously. This is
+similar to how many implementations do HTTP request/response:
+
+- JS [Fetch](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) api returns a [`Response`](https://fetch.spec.whatwg.org/#response) object as soon as the server returns headers. The
+  body of both requests and responses are `ReadableStreams`.
+- Go's [http.Request](https://pkg.go.dev/net/http@go1.20.5#Request) and [http.Response](https://pkg.go.dev/net/http@go1.20.5#Response) bodies are both [`io.ReadCloser`](https://pkg.go.dev/io@go1.20.5#ReadCloser) types.
+- The Rust [hyper](https://docs.rs/hyper) crate has a similar [`HTTPBody`](https://docs.rs/hyper/latest/hyper/body/trait.HttpBody.html) that reads request/response data
+  incrementally and asynchronously.
+
+## Goals of this document
+
+The primary goal of this document is to define an interface that application
+protocols can build on and transports can fulfill.
+
+Another goal is to be backwards compatible with existing application protocols
+that follow these semantics to make upgrading them easier so that they may take
+advantage of the new HTTP transport without sacrificing backwards compatibility.
+
+The final goal is to define something simple such that it's easy to follow,
+implement, and use.
+
+## How to map to a libp2p stream
+
+Each request and response should happen in a single stream. There SHOULD NOT be
```

Comment:
```suggestion
Each request and response should happen in a single stream. There MUST NOT be
```

Reply:
For backwards compatibility, this can't be a MUST.

I'd like to have existing request/response over streams (like identify), to already be compliant with this spec.

Reply:
Previous discussion: https://github.com/libp2p/specs/pull/561#discussion_r1265018181